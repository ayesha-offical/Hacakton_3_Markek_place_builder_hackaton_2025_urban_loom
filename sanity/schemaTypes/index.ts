import { type SchemaTypeDefinition } from 'sanity'

import {categoryType} from './categoryType'

import { salesService } from './sale'
import { productType } from './ProductType'

export const schema: { types: SchemaTypeDefinition[] } = {
  types: [ categoryType, salesService,productType],
}