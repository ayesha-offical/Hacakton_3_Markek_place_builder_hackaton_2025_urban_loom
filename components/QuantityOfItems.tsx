import { Product } from "@/sanity.types";
import { Button } from "./ui/button";
import {HiMinus, HiPlus} from "react-icons/hi2";
import toast from "react-hot-toast";
import { cn } from "@/lib/utils";

interface Props {
  product: Product;
  className?: string;
  borderStyle?: string;
}

const QuantityOfItems = ({product,className,borderStyle}:Props) => {
    const handleRemoveItem = () => {
        // implement logic to remove item from the cart
        toast.success("Item removed")
    };
    const handleAddItem = () => {
        // implement logic to add item to the cart
        toast.success("Item added")
    };
    const itemCount=5;
  return (
    <div className={cn('flex items-center gap-1 pb-1 text-base')}>
        <Button variant="outline" size="icon" className="w-6 h-6"onClick={handleRemoveItem}>
            <HiMinus/>
        </Button>

    <span className="font-semibold w-5 text-center text-slate-800">{itemCount}</span>
    <Button variant="outline" size="icon" className="w-6 h-6"onClick={handleAddItem}>
            <HiPlus/>
        </Button>
    </div>
  )
}

export default QuantityOfItems 