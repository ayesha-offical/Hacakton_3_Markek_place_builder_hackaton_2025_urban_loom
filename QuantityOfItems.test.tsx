// import '@testing-library/jest-dom'
// import { render, screen } from '@testing-library/react'
// import QuantityOfItems from './components/QuantityOfItems'

// describe('QuantityOfItems', () => {
//   const mockProduct = {
//     _id: '123',
//     name: 'Test Product',
//     price: 100,
//     stock: 10,
//     slug: { current: 'test-product' }
//   }

//   it('renders quantity controls', () => {
//     render(<QuantityOfItems product={mockProduct} />)
    
//     const decreaseButton = screen.getByRole('button', { name: /decrease quantity/i })
//     const increaseButton = screen.getByRole('button', { name: /increase quantity/i })
//     const quantityDisplay = screen.getByText('0')

//     expect(decreaseButton).toBeInTheDocument()
//     expect(increaseButton).toBeInTheDocument()
//     expect(quantityDisplay).toBeInTheDocument()
//   })

//   it('renders correct button icons', () => {
//     render(<QuantityOfItems product={mockProduct} />)
    
//     const decreaseButton = screen.getByRole('button', { name: /decrease quantity/i })
//     const increaseButton = screen.getByRole('button', { name: /increase quantity/i })

//     expect(decreaseButton.querySelector('svg')).toBeInTheDocument()
//     expect(increaseButton.querySelector('svg')).toBeInTheDocument()
//   })

//   it('has proper accessibility labels', () => {
//     render(<QuantityOfItems product={mockProduct} />)
    
//     expect(screen.getByLabelText(/decrease quantity/i)).toBeInTheDocument()
//     expect(screen.getByLabelText(/increase quantity/i)).toBeInTheDocument()
//   })
// })


import { render, screen, fireEvent } from "@testing-library/react";
import QuantityOfItems from "@/components/QuantityOfItems"; // Adjust import path if necessary
import useCartStore from "./store";
import toast from "react-hot-toast";
import { Product } from "@/sanity.types";
import "@testing-library/jest-dom";

// Mock the Zustand store
jest.mock("./store", () => ({
  __esModule: true,
  default: jest.fn(),
}));

// Mock react-hot-toast
jest.mock("react-hot-toast", () => ({
  success: jest.fn(),
}));

// Sample product for testing
const mockProduct: Product = {
  _id: "prod-1",
  name: "Test Product",
  price: 100,
  stock: 10,
  description: "Test Description",
} as Required<Pick<Product, 'name'>> & Product;

describe("QuantityOfItems Component", () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  test("renders correctly with the given quantity", () => {
    // Mock store to return a specific quantity
    (useCartStore as unknown as jest.Mock).mockReturnValue({
      addItem: jest.fn(),
      removeItem: jest.fn(),
      getItemCount: jest.fn(() => 3),
    });

    render(<QuantityOfItems product={mockProduct} />);

    // Ensure the displayed quantity is correct
    expect(screen.getByText("3")).toBeInTheDocument();
    // Check that both increase and decrease buttons are rendered
    expect(screen.getByRole("button", { name: /Decrease Quantity/i })).toBeInTheDocument();
    expect(screen.getByRole("button", { name: /Increase Quantity/i })).toBeInTheDocument();
  });

  test("calls removeItem and shows 'Quantity Decreased' toast when quantity is greater than 1", () => {
    const mockRemoveItem = jest.fn();
    (useCartStore as unknown as jest.Mock).mockReturnValue({
      addItem: jest.fn(),
      removeItem: mockRemoveItem,
      getItemCount: jest.fn(() => 2), // More than 1 item
    });

    render(<QuantityOfItems product={mockProduct} />);

    const decreaseButton = screen.getByRole("button", { name: /Decrease Quantity/i });
    fireEvent.click(decreaseButton);

    // Ensure removeItem is called with the product ID
    expect(mockRemoveItem).toHaveBeenCalledWith(mockProduct._id);
    // Toast should show "Quantity Decreased"
    expect(toast.success).toHaveBeenCalledWith("Quantity Decreased");
  });

  test("calls removeItem and shows removed message when quantity is 1 or less", () => {
    const mockRemoveItem = jest.fn();
    (useCartStore as unknown as jest.Mock).mockReturnValue({
      addItem: jest.fn(),
      removeItem: mockRemoveItem,
      getItemCount: jest.fn(() => 1), // 1 item or less
    });

    render(<QuantityOfItems product={mockProduct} />);

    const decreaseButton = screen.getByRole("button", { name: /Decrease Quantity/i });
    fireEvent.click(decreaseButton);

    expect(mockRemoveItem).toHaveBeenCalledWith(mockProduct._id);

    // Toast should show a message with product name substring and removed text
    const expectedMessage = `${(mockProduct.name ?? "Product").substring(0, 12)}....removed successfully`;
    expect(toast.success).toHaveBeenCalledWith(expectedMessage);
  });

  test("calls addItem and shows 'Quantity Increased' toast when increase button is clicked", () => {
    const mockAddItem = jest.fn();
    (useCartStore as unknown as jest.Mock).mockReturnValue({
      addItem: mockAddItem,
      removeItem: jest.fn(),
      getItemCount: jest.fn(() => 1),
    });

    render(<QuantityOfItems product={mockProduct} />);

    const increaseButton = screen.getByRole("button", { name: /Increase Quantity/i });
    fireEvent.click(increaseButton);

    // Ensure addItem is called with the product object
    expect(mockAddItem).toHaveBeenCalledWith(mockProduct);
    // Toast should show "Quantity Increased"
    expect(toast.success).toHaveBeenCalledWith("Quantity Increased");
  });
  test("matches snapshot", () => {
    (useCartStore as unknown as jest.Mock).mockReturnValue({
      addItem: jest.fn(),
      removeItem: jest.fn(),
      getItemCount: jest.fn(() => 5),
    });

    const { container } = render(<QuantityOfItems product={mockProduct} />);
    expect(container).toMatchSnapshot();
  });
});
