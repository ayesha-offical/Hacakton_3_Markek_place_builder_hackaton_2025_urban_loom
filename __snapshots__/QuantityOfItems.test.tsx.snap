// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuantityOfItems Component matches snapshot 1`] = `
<div>
  <div
    class="flex items-center gap-1 pb-1 text-base"
  >
    <button
      aria-label="Decrease Quantity"
      class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 border border-input bg-background shadow-sm hover:bg-accent hover:text-accent-foreground w-6 h-6"
    >
      <svg
        aria-hidden="true"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M4.25 12a.75.75 0 0 1 .75-.75h14a.75.75 0 0 1 0 1.5H5a.75.75 0 0 1-.75-.75Z"
          fill-rule="evenodd"
        />
      </svg>
    </button>
    <span
      class="font-semibold w-5 text-center text-slate-800"
    >
      5
    </span>
    <button
      aria-label="Increase Quantity"
      class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 border border-input bg-background shadow-sm hover:bg-accent hover:text-accent-foreground w-6 h-6"
    >
      <svg
        aria-hidden="true"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M12 3.75a.75.75 0 0 1 .75.75v6.75h6.75a.75.75 0 0 1 0 1.5h-6.75v6.75a.75.75 0 0 1-1.5 0v-6.75H4.5a.75.75 0 0 1 0-1.5h6.75V4.5a.75.75 0 0 1 .75-.75Z"
          fill-rule="evenodd"
        />
      </svg>
    </button>
  </div>
</div>
`;
